### PR TITLE
Navi Search Load Search Results Individually

### PR DESCRIPTION
**Pull request by ldkge**
Resolves #

## Description
It used to be that for the navi-search-provider federated service to return results it had to successfully resolve the responses of all the search providers. 

Now changed so that the federated service returns tasks instances of the search providers and the result wrapper is the one that takes care of displaying only the results that successfully resolved from the task instances.

## Proposed Changes

-

## Screenshots
Screen Recording 2020-07-22 at 3 49 18 PM

## License

I confirm that this contribution is made under an MIT license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: packages/search/tests/unit/services/navi-search-provider-test.js [hunk 1]
```diff
@@ -42,19 +42,13 @@ module('Unit | Service | navi-search-provider', function(hooks) {
...
-    let expectedResults = ['Revenue result', 'Revenue success'];
+    let availableSearchProviders = service._all();
+    let results = service.search('sample');
+
     assert.deepEqual(
-      results.find(result => result.component === 'navi-search-result/sample').data,
-      expectedResults,
-      'Search returns the expected results'
+      results.map(result => result.context.constructor.name),
+      availableSearchProviders.map(result => result.constructor.name),
```
Comment (kevinhinterlong) [inline packages/search/tests/unit/services/navi-search-provider-test.js:50]: use an explicit list here

File: packages/search/addon/components/navi-search-result/wrapper.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+interface Args {
+  searchResults: Array<TODO>;
+}
+
+export default class NaviWrapperSearchResultComponent extends Component<Args> {
+  /**
+   * @property {boolean} areTasksRunning
+   */
+  get areTasksRunning() {
+    return this.args?.searchResults.some(searchResult => searchResult.isRunning);
```
Comment (kevinhinterlong) [inline packages/search/addon/components/navi-search-result/wrapper.ts:22]:
```suggestion
    return this.args.searchResults.some(searchResult => searchResult.isRunning);
```

File: packages/search/addon/components/navi-search-result/wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  }
+
+  /**
+   * @property {boolean} areTasksRunning
+   */
+  get areResultsEmpty() {
+    return this.args?.searchResults.every(searchResult => {
```
Comment (kevinhinterlong) [inline packages/search/addon/components/navi-search-result/wrapper.ts:29]:
```suggestion
    return this.args.searchResults.every(searchResult => {
```

File: packages/search/addon/components/navi-search-result/wrapper.hbs [hunk 4]
```diff
@@ -1,9 +1,23 @@
...
+    {{/let}}
+  {{/if}}
+{{/each}}
+{{#if this.areResultsEmpty}}
   <li class="navi-search-result-wrapper__item" ...attributes>
-    {{#if result.title}}
-      <h3 class="navi-search-result-wrapper__item-title">{{result.title}}</h3>
-    {{/if}}
-    {{component result.component data=result.data closeResults=@closeResults}}
+    <NaviSearchResult::NoResult closeResults={{@closeResults}} />
```
Comment (kevinhinterlong) [inline packages/search/addon/components/navi-search-result/wrapper.hbs:18]:
```suggestion
    <NaviSearchResult::NoResult @closeResults={{@closeResults}} />
```

File: packages/search/addon/components/navi-search-result/wrapper.hbs [hunk 5]
```diff
@@ -1,9 +1,23 @@
...
-{{#each @searchResults as |result|}}
+{{#each @searchResults as |searchResult|}}
+  {{#if searchResult.isSuccessful}}
+    {{#let searchResult.value as |result|}}
+      {{#if (gt result.data.length 0)}}
+        <li class="navi-search-result-wrapper__item" ...attributes>
+          {{#if result.title}}
+            <h3 class="navi-search-result-wrapper__item-title">{{result.title}}</h3>
+          {{/if}}
+          {{component result.component data=result.data closeResults=@closeResults}}
```
Comment (kevinhinterlong) [inline packages/search/addon/components/navi-search-result/wrapper.hbs:10]:
```suggestion
          {{#let (component result.component) as |SearchResult|}}
            <SearchResult @data={{result.data}} @closeResults={{@closeResults}} />
          {{/let}}
```

File: packages/search/addon/components/navi-search-result/wrapper.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+export default class NaviWrapperSearchResultComponent extends Component<Args> {
+  /**
+   * @property {boolean} areTasksRunning
+   */
+  get areTasksRunning() {
+    return this.args?.searchResults.some(searchResult => searchResult.isRunning);
+  }
+
+  /**
+   * @property {boolean} areTasksRunning
```
Comment (kevinhinterlong) [inline packages/search/addon/components/navi-search-result/wrapper.ts:26]:
```suggestion
   * @property {boolean} areResultsEmpty
```

File: packages/search/app/styles/navi-search/components/navi-wrapper-search-result.less [hunk 7]
```diff
@@ -19,4 +19,12 @@
     line-height: 1.5;
     margin-block-start: 0;
   }
+
+  &__loader {
+    height: 10px;
```
Comment (kevinhinterlong) [inline packages/search/app/styles/navi-search/components/navi-wrapper-search-result.less:24]:
```suggestion
    height: 10px;
    
    & > .navi-loader__spinner {
      background-color: @denali-grey-100;
    }
```

File: packages/search/app/styles/navi-search/components/navi-wrapper-search-result.less [hunk 8]
```diff
@@ -19,4 +19,12 @@
...
+  }
+}
+
+.navi-search-result-wrapper__loader > .navi-loader__spinner {
+  background-color: @denali-grey-100;
 }
```
Comment (kevinhinterlong) [inline packages/search/app/styles/navi-search/components/navi-wrapper-search-result.less:30]:
```suggestion
```

File: packages/search/tests/integration/components/navi-search-result/wrapper-test.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+      hbs`<NaviSearchResult::Wrapper @searchResults={{this.searchResults}} @closeResults={{this.closeResults}} />`
+    );
+
+    assert
+      .dom('.navi-search-result-wrapper__item-title')
+      .hasText('Sample Loaded', 'Loaded search results are displayed');
+    assert.dom('.navi-search-result-wrapper__loader').exists('Loader is shown for pending search results.');
+  });
+
+  test('results rendered', async function(assert) {
```
Comment (kevinhinterlong) [inline packages/search/tests/integration/components/navi-search-result/wrapper-test.js:114]: Maybe add one more test showing when 2 search providers finished with data

File: packages/search/tests/integration/components/navi-search-result/wrapper-test.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+        }
+      }
+    ]);
+    set(this, 'closeResults', {});
+
+    await render(
+      hbs`<NaviSearchResult::Wrapper @searchResults={{this.searchResults}} @closeResults={{this.closeResults}} />`
+    );
+
+    assert.dom('.navi-search-no-result').exists('No Results component shown');
```
Comment (kevinhinterlong) [inline packages/search/tests/integration/components/navi-search-result/wrapper-test.js:30]: assert loader is not shown as well

File: packages/search/tests/integration/components/navi-search-result/wrapper-test.js [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
```
Comment (kevinhinterlong) [inline packages/search/tests/integration/components/navi-search-result/wrapper-test.js:53]: assert loader is not shown as well

File: packages/search/tests/integration/components/navi-search-result/wrapper-test.js [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        }
+      }
+    ]);
+    set(this, 'closeResults', {});
+
+    await render(
+      hbs`<NaviSearchResult::Wrapper @searchResults={{this.searchResults}} @closeResults={{this.closeResults}} />`
+    );
+
+    assert.dom('.navi-search-result-wrapper__loader').exists('Loader is shown.');
```
Comment (kevinhinterlong) [inline packages/search/tests/integration/components/navi-search-result/wrapper-test.js:76]: assert no results is not shown as well

File: packages/search/tests/integration/components/navi-search-result/wrapper-test.js [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+    set(this, 'closeResults', {});
+
+    await render(
+      hbs`<NaviSearchResult::Wrapper @searchResults={{this.searchResults}} @closeResults={{this.closeResults}} />`
+    );
+
+    assert
+      .dom('.navi-search-result-wrapper__item-title')
+      .hasText('Sample Loaded', 'Loaded search results are displayed');
+    assert.dom('.navi-search-result-wrapper__loader').exists('Loader is shown for pending search results.');
```
Comment (kevinhinterlong) [inline packages/search/tests/integration/components/navi-search-result/wrapper-test.js:111]: assert no results is not shown as well

File: packages/search/tests/integration/components/navi-search-result/wrapper-test.js [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    ]);
+    set(this, 'closeResults', {});
+
+    await render(
+      hbs`<NaviSearchResult::Wrapper @searchResults={{this.searchResults}} @closeResults={{this.closeResults}} />`
+    );
+
+    assert
+      .dom('.navi-search-result-wrapper__item-title')
+      .hasText('Sample Loaded', 'Loaded search results are displayed');
```
Comment (kevinhinterlong) [inline packages/search/tests/integration/components/navi-search-result/wrapper-test.js:136]: assert loader is not shown as well

File: packages/search/addon/components/navi-search-result/wrapper.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+  /**
+   * @property {boolean} areResultsEmpty
+   */
+  get areResultsEmpty() {
+    return this.args.searchResults.every(searchResult => {
+      return (
+        (!searchResult.isRunning && !searchResult.isSuccessful) ||
+        (!searchResult.isRunning && !(searchResult.value.data.length > 0))
+      );
```
Comment (mattgraham1995) [inline packages/search/addon/components/navi-search-result/wrapper.ts:33]:
```suggestion
      return !searchResult.isRunning && (!searchResult.isSuccessful || !searchResult.value.data.length);
```

File: packages/search/tests/integration/components/navi-search-result/wrapper-test.js [hunk 16]
```diff
@@ -0,0 +1,180 @@
+import { module, test } from 'qunit';
+import { setupRenderingTest } from 'ember-qunit';
+import { render, findAll } from '@ember/test-helpers';
+import { hbs } from 'ember-cli-htmlbars';
+import { set } from '@ember/object';
+
+module('Integration | Component | navi-search-result-wrapper', function(hooks) {
+  setupRenderingTest(hooks);
+
+  test('no results tasks are not succeeded', async function(assert) {
```
Comment (mattgraham1995) [inline packages/search/tests/integration/components/navi-search-result/wrapper-test.js:10]:
```suggestion
  test('no successful results tasks', async function(assert) {
```

File: packages/search/tests/integration/components/navi-search-result/wrapper-test.js [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+
+    await render(
+      hbs`<NaviSearchResult::Wrapper @searchResults={{this.searchResults}} @closeResults={{this.closeResults}} />`
+    );
+
+    assert.dom('.navi-search-no-result').exists('No Results component shown');
+    assert.dom('.navi-search-result-wrapper__loader').doesNotExist('Loader is not shown');
+  });
+
+  test('no results tasks return no data', async function(assert) {
```
Comment (mattgraham1995) [inline packages/search/tests/integration/components/navi-search-result/wrapper-test.js:34]:
```suggestion
  test('no results tasks return data', async function(assert) {
```

**Pull request merged**